File: app/src/main/java/com/arjanvlek/oxygenupdater/views/DeviceInformationFragment.java
package com.arjanvlek.oxygenupdater.views;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.arjanvlek.oxygenupdater.Model.Device;
import com.arjanvlek.oxygenupdater.Model.DeviceInformationData;
import com.arjanvlek.oxygenupdater.Model.SystemVersionProperties;
import com.arjanvlek.oxygenupdater.R;

import java.util.List;

import java8.util.stream.StreamSupport;

import static com.arjanvlek.oxygenupdater.ApplicationContext.NO_OXYGEN_OS;

public class DeviceInformationFragment extends AbstractFragment {
    private RelativeLayout rootView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        //Inflate the layout for this fragment
        rootView = (RelativeLayout) inflater.inflate(R.layout.fragment_deviceinformation, container, false);
        return rootView;
    }


    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        if(isAdded()) {
            displayDeviceInformation();
            getApplicationContext().getDevices(this::displayDeviceName);
        }
    }

    private void displayDeviceName(List<Device> devices) {

        if(isAdded()) {
            TextView deviceNameView = (TextView) rootView.findViewById(R.id.device_information_header);
            SystemVersionProperties systemVersionProperties = getApplicationContext().getSystemVersionProperties();

            if (devices != null) {
                StreamSupport.stream(devices)
                        .filter(device -> device.getProductName() != null && device.getProductName().equals(systemVersionProperties.getOxygenDeviceName()))
                        .forEach(device -> deviceNameView.setText(device.getName()));
            }
        }
    }

    private void displayDeviceInformation() {
        if(isAdded()) {
            DeviceInformationData deviceInformationData = new DeviceInformationData();
            SystemVersionProperties systemVersionProperties = getApplicationContext().getSystemVersionProperties();

            TextView deviceNameView = (TextView) rootView.findViewById(R.id.device_information_header);
            deviceNameView.setText(String.format(getString(R.string.device_information_device_name), deviceInformationData.getDeviceManufacturer(), deviceInformationData.getDeviceName()));

            TextView socView = (TextView) rootView.findViewById(R.id.device_information_soc_field);
            socView.setText(deviceInformationData.getSoc());

            String cpuFreqString = deviceInformationData.getCpuFrequency();
            TextView cpuFreqView = (TextView) rootView.findViewById(R.id.device_information_cpu_freq_field);
            if (!cpuFreqString.equals(DeviceInformationData.UNKNOWN)) {
                cpuFreqView.setText(String.format(getString(R.string.device_information_gigahertz), deviceInformationData.getCpuFrequency()));
            } else {
                cpuFreqView.setText(getString(R.string.device_information_unknown));
            }

            long totalMemory;
            try {
                ActivityManager.MemoryInfo mi = new ActivityManager.MemoryInfo();
                ActivityManager activityManager = (ActivityManager) getActivity().getBaseContext().getSystemService(Context.ACTIVITY_SERVICE);
                activityManager.getMemoryInfo(mi);
                totalMemory = mi.totalMem / 1048576L;
            } catch (Exception ignored) {
                totalMemory = 0;
            }

            TextView memoryView = (TextView) rootView.findViewById(R.id.device_information_memory_field);
            if (totalMemory != 0) {
                memoryView.setText(String.format(getString(R.string.download_size_megabyte), totalMemory));
            } else {
                View memoryLabel = rootView.findViewById(R.id.device_information_memory_label);
                memoryLabel.setVisibility(View.GONE);
                memoryView.setVisibility(View.GONE);
            }

            TextView oxygenOsVersionView = (TextView) rootView.findViewById(R.id.device_information_oxygen_os_ver_field);

            if (!systemVersionProperties.getOxygenOSVersion().equals(NO_OXYGEN_OS)) {
                oxygenOsVersionView.setText(systemVersionProperties.getOxygenOSVersion());

            } else {
                TextView oxygenOsVersionLabel = (TextView) rootView.findViewById(R.id.device_information_oxygen_os_ver_label);
                oxygenOsVersionLabel.setVisibility(View.GONE);
                oxygenOsVersionView.setVisibility(View.GONE);
            }

            TextView osVerView = (TextView) rootView.findViewById(R.id.device_information_os_ver_field);
            osVerView.setText(deviceInformationData.getOsVersion());

            TextView osIncrementalView = (TextView) rootView.findViewById(R.id.device_information_incremental_os_ver_field);
            osIncrementalView.setText(deviceInformationData.getIncrementalOsVersion());

            TextView osPatchDateView = (TextView) rootView.findViewById(R.id.device_information_os_patch_level_field);

            if (!systemVersionProperties.getSecurityPatchDate().equals(NO_OXYGEN_OS)) {
                osPatchDateView.setText(systemVersionProperties.getSecurityPatchDate());
            } else {
                TextView osPatchDateLabel = (TextView) rootView.findViewById(R.id.device_information_os_patch_level_label);
                osPatchDateLabel.setVisibility(View.GONE);
                osPatchDateView.setVisibility(View.GONE);
            }

            TextView serialNumberView = (TextView) rootView.findViewById(R.id.device_information_serial_number_field);
            serialNumberView.setText(deviceInformationData.getSerialNumber());
        }
    }

}

File: app/src/main/java/com/arjanvlek/oxygenupdater/fragments/DeviceInformationFragment.kt
package com.arjanvlek.oxygenupdater.fragments

import android.app.ActivityManager
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import com.arjanvlek.oxygenupdater.OxygenUpdater.Companion.NO_OXYGEN_OS
import com.arjanvlek.oxygenupdater.R
import com.arjanvlek.oxygenupdater.activities.MainActivity
import com.arjanvlek.oxygenupdater.internal.DeviceInformationData
import com.arjanvlek.oxygenupdater.models.Device
import com.arjanvlek.oxygenupdater.models.DeviceOsSpec
import com.arjanvlek.oxygenupdater.utils.Logger
import com.arjanvlek.oxygenupdater.utils.Logger.logDebug
import com.arjanvlek.oxygenupdater.utils.Utils
import com.arjanvlek.oxygenupdater.viewmodels.MainViewModel
import kotlinx.android.synthetic.main.fragment_device_information.*
import org.koin.androidx.viewmodel.ext.android.sharedViewModel

class DeviceInformationFragment : AbstractFragment() {

    private val mainViewModel: MainViewModel by sharedViewModel()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = super.onCreateView(inflater, container, savedInstanceState).let {
        inflater.inflate(R.layout.fragment_device_information, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        if (!isAdded) {
            logDebug(TAG, "Fragment not added. Can not create the view!")
            return
        }

        displayDeviceInformation()

        mainViewModel.allDevices.observe(viewLifecycleOwner, Observer { devices ->
            updateBannerText(Utils.checkDeviceOsSpec(application!!.systemVersionProperties!!, devices))

            displayFormattedDeviceName(devices)
        })
    }

    private fun updateBannerText(deviceOsSpec: DeviceOsSpec) {
        bannerTextView.text = getString(
            when (deviceOsSpec) {
                DeviceOsSpec.SUPPORTED_OXYGEN_OS -> R.string.device_information_supported_oxygen_os
                DeviceOsSpec.CARRIER_EXCLUSIVE_OXYGEN_OS -> R.string.device_information_carrier_exclusive_oxygen_os
                DeviceOsSpec.UNSUPPORTED_OXYGEN_OS -> R.string.device_information_unsupported_oxygen_os
                DeviceOsSpec.UNSUPPORTED_OS -> R.string.device_information_unsupported_os
            }
        )

        bannerLayout.isVisible = true

        if (!deviceOsSpec.isDeviceOsSpecSupported) {
            bannerLayout.setOnClickListener { (activity as MainActivity?)?.displayUnsupportedDeviceOsSpecMessage(deviceOsSpec) }
        }
    }

    private fun displayFormattedDeviceName(devices: List<Device>?) {
        if (!isAdded) {
            logDebug(TAG, "Fragment not added. Can not display formatted device name!")
            return
        }

        devices?.find { device -> device.productNames.contains(application?.systemVersionProperties?.oxygenDeviceName) }
            ?.let { device -> device_information_header.text = device.name }
    }

    private fun displayDeviceInformation() {
        if (!isAdded) {
            logDebug(TAG, "Fragment not added. Can not display device information!")
            return
        }

        val systemVersionProperties = application?.systemVersionProperties

        // Device name (in top)
        device_information_header.text =
            getString(R.string.device_information_device_name, DeviceInformationData.deviceManufacturer, DeviceInformationData.deviceName)

        // SoC
        device_information_soc_field.text = DeviceInformationData.soc

        // CPU Frequency (if available)
        device_information_cpu_freq_field.apply {
            val cpuFreqString = DeviceInformationData.cpuFrequency

            text = if (cpuFreqString != DeviceInformationData.UNKNOWN) {
                getString(R.string.device_information_gigahertz, cpuFreqString)
            } else {
                getString(R.string.device_information_unknown)
            }
        }

        val totalMemory: Long = try {
            val mi = ActivityManager.MemoryInfo()
            val activityManager = Utils.getSystemService(context, Context.ACTIVITY_SERVICE) as ActivityManager

            activityManager.getMemoryInfo(mi)
            mi.totalMem / 1048576L
        } catch (e: Exception) {
            Logger.logWarning("DeviceInformationFragment", "Memory information is unavailable due to error", e)
            0
        }

        // Total amount of RAM (if available)
        device_information_memory_field.apply {
            if (totalMemory != 0L) {
                text = getString(R.string.download_size_megabyte, totalMemory)
            } else {
                device_information_memory_label.isVisible = false
                isVisible = false
            }
        }

        // OxygenOS version (if available)
        device_information_oxygen_os_ver_field.apply {
            val oxygenOSVersion = systemVersionProperties?.oxygenOSVersion

            if (oxygenOSVersion != NO_OXYGEN_OS) {
                text = oxygenOSVersion
            } else {
                device_information_oxygen_os_ver_label.isVisible = false
                isVisible = false
            }
        }

        // OxygenOS OTA version (if available)
        device_information_oxygen_os_ota_ver_field.apply {
            val oxygenOSOTAVersion = systemVersionProperties?.oxygenOSOTAVersion

            if (oxygenOSOTAVersion != NO_OXYGEN_OS) {
                text = oxygenOSOTAVersion
            } else {
                isVisible = false
            }
        }

        // Android version
        device_information_os_ver_field.text = DeviceInformationData.osVersion

        // Incremental OS version
        device_information_incremental_os_ver_field.text = DeviceInformationData.incrementalOsVersion

        // Security Patch Date (if available)
        device_information_os_patch_level_field.apply {
            val securityPatchDate = systemVersionProperties?.securityPatchDate

            if (securityPatchDate != NO_OXYGEN_OS) {
                text = securityPatchDate
            } else {
                device_information_os_patch_level_label.isVisible = false
                isVisible = false
            }
        }

        // Serial number (Android 7.1.2 and lower only)
        device_information_serial_number_field.apply {
            val serialNumber = DeviceInformationData.serialNumber

            if (serialNumber != DeviceInformationData.UNKNOWN) {
                text = serialNumber
            } else {
                device_information_serial_number_label.isVisible = false
                isVisible = false
            }
        }
    }

    companion object {
        private const val TAG = "DeviceInformationFragment"
    }
}
